cli/cmd/root.go

```diff
@@ -189,6 +191,27 @@ func getSuccessRate(success, failure uint64) float64 {
 	return float64(success) / float64(success+failure)
 }
 
+// getDefaultNamespace fetches the default namespace
+// used in the current KubeConfig context
+func getDefaultNamespace() string {
+	rules := clientcmd.NewDefaultClientConfigLoadingRules()
+
+	if kubeconfigPath != "" {
+		rules.ExplicitPath = kubeconfigPath
+	}
+
+	overrides := &clientcmd.ConfigOverrides{CurrentContext: kubeContext}
+	kubeCfg := clientcmd.NewNonInteractiveDeferredLoadingClientConfig(rules, overrides)
+	ns, _, err := kubeCfg.Namespace()
+
+	if err != nil {
+		log.Errorf("failed to set namespace from config context:%v", err)
+		return defaultContextNamespace
+	}
```

Comment:
I start to wonder how severe this error actually is. Do we want to recover from it or we rather fail the whole command invocation right away? Is it safe to assume that if we error out here, we can simply return the "default" namespace ?

Reply:
Hi @zaharidichev  I've looked at the `Namespace()` method implementation in the `DeferredLoadingClientConfig` type to determine what will explicitly happen if an error is encountered.  The function errors out if the client config cannot be created properly (in which case it returns an empty string), the client config is created based on the rules and rule loader that are created at the start of the `getDefaultNamespace()` function. The rest of the `Namespace()`method either returns the namespace defined in the context or if the namespace is `"default"` tries to work out whether it was an implicit or explicit choice.

> Is it safe to assume that if we error out here, we can simply return the "default" namespace

I assume so, yes. If the function fails, it means for whatever reason, the config was not properly initialised/created. The single point of error when creating the client config is when the rules are loaded in. A reason for failure, in this case, would be that the config cannot be parsed or loaded from a `kubeConfig` file.

To my understanding (and skill), this is what is happening under the hood when we are creating and loading those config "rules". Based on this, I think on error it is safe to just return the `"default"` namespace. In this context (no pun intended) it seems a default is more sensible since we might not know what the underlying issues with the `kubeConfig` are in the `cli` client environment.

Tl;dr To my understanding the point of error happens when either the config or the rules are not loaded in properly, in which case I think it is sensible to just return a default. If the problem persists (i.e context namespace simply won't be inferred), the namespace can just be overridden with the `-n` flag. If instead we just fail the command invocation outright, it might be that none of the commands will work, since all `cli` command options are making a call to `getDefaultNamespace()` so options will never be initialised if it all fails.

Of course, this is just my opinion, look forward to hearing what you think :)

Reply:
Yes, I see. That makes sense to me.